tests/InstrEngineTests/NaglerInstrumentationMethod/dllmain.cpp

```diff
@@ -5,17 +5,79 @@
 
 #include "stdafx.h"
 #include "resource.h"
-#include "dllmain.h"
+#include "NaglerInstrumentationMethod.h"
 
-CInstrumentationMethodModule _AtlModule;
+class CCrossPlatClassFactory : public IClassFactory, public CModuleRefCount
+{
+private:
+
+public:
+    DEFINE_DELEGATED_REFCOUNT_ADDREF(CCrossPlatClassFactory);
+    DEFINE_DELEGATED_REFCOUNT_RELEASE(CCrossPlatClassFactory);
+    STDMETHOD(STDMETHODCALLTYPE) QueryInterface(REFIID riid, PVOID* ppvObject) override
+    {
+        HRESULT hr = E_NOINTERFACE;
+
+        hr = ImplQueryInterface(
+            static_cast<IClassFactory*>(this),
+            riid, ppvObject
+        );
+
+        return hr;
+    }
+
+    // IClassFactory Methods
+public:
+    STDMETHOD(CreateInstance)(
+        _In_ IUnknown* pUnkOuter,
+        _In_   REFIID   riid,
+        _Out_ void** ppvObject
+        ) override
+    {
+        HRESULT hr = S_OK;
+
+        CComPtr<CInstrumentationMethod> pInstrumentationMethod;
+        pInstrumentationMethod.Attach(new CInstrumentationMethod);
+        if (pInstrumentationMethod == NULL)
+        {
+            return E_OUTOFMEMORY;
+        }
+
+        IfFailRet(pInstrumentationMethod->QueryInterface(riid, ppvObject));
 
-// DLL Entry Point
-extern "C" BOOL WINAPI DllMain(HINSTANCE hInstance, DWORD dwReason, LPVOID lpReserved)
+        return S_OK;
+    }
+
+    STDMETHOD(LockServer)(
+        _In_ BOOL fLock
+        ) override
+    {
+        return E_NOTIMPL;
+    }
+};
+
+// Used to determine whether the DLL can be unloaded by OLE.
+__control_entrypoint(DllExport)
+STDAPI DLLEXPORT(DllCanUnloadNow, 0)(void)
 {
-#ifdef _MERGE_PROXYSTUB
-    if (!PrxDllMain(hInstance, dwReason, lpReserved))
-        return FALSE;
-#endif
-    hInstance;
-    return _AtlModule.DllMain(dwReason, lpReserved);
+    if (CModuleRefCount::GetModuleUsage() == 0)
+    {
+        return S_OK;
+    }
+
+    return S_FALSE;
 }
+
+// Returns a class factory to create an object of the requested type.
+_Check_return_
+STDAPI DLLEXPORT(DllGetClassObject, 12)(_In_ REFCLSID rclsid, _In_ REFIID riid, _Outptr_ LPVOID FAR* ppv)
```

Comment:
>DLLEXPORT [](http://example.com/codeflow?start=7&length=9)

Since you were not able to do it in this PR, coudl you add an issue to GitHub for exporting DllGetClassObject, etc. in Nagler for Linux?